modal/cli/import_refs.py

```diff
@@ -92,136 +93,149 @@ def import_file_or_module(file_or_module: str):
     return module
 
 
-@dataclass
+@dataclass(frozen=True)
 class MethodReference:
     """This helps with deferring method reference until after the class gets instantiated by the CLI"""
 
     cls: Cls
     method_name: str
 
 
-def get_by_object_path(obj: Any, obj_path: str) -> Union[Function, LocalEntrypoint, MethodReference, App, None]:
-    # Try to evaluate a `.`-delimited object path in a Modal context
-    # With the caveat that some object names can actually have `.` in their name (lifecycled methods' tags)
-
-    # Note: this is eager, so no backtracking is performed in case an
-    # earlier match fails at some later point in the path expansion
-    prefix = ""
-    obj_path_segments = obj_path.split(".")
-    for i, segment in enumerate(obj_path_segments):
-        attr = prefix + segment
-        if isinstance(obj, App):
-            if attr in obj.registered_entrypoints:
-                # local entrypoints can't be accessed via getattr
-                obj = obj.registered_entrypoints[attr]
-                continue
-        if isinstance(obj, Cls):
-            remaining_segments = obj_path_segments[i:]
-            remaining_path = ".".join(remaining_segments)
-            if len(remaining_segments) > 1:
-                raise ValueError(f"{obj._get_name()} is a class, but {remaining_path} is not a method reference")
-            # TODO: add method inference here?
-            return MethodReference(obj, remaining_path)
+Runnable = Union[Function, MethodReference, LocalEntrypoint]
 
-        try:
-            obj = getattr(obj, attr)
-        except Exception:
-            prefix = f"{prefix}{segment}."
-        else:
-            prefix = ""
 
-    if prefix:
-        return None
+@dataclass(frozen=True)
+class CLICommand:
+    names: list[str]
+    runnable: Runnable
+    is_web_endpoint: bool
 
-    return obj
 
+def list_cli_commands(
+    module: types.ModuleType,
+) -> list[CLICommand]:
+    """
+    Extracts all runnables found either directly in the input module, or in any of the Apps listed in that module
 
-def _infer_function_or_help(
-    app: App, module, accept_local_entrypoint: bool, accept_webhook: bool
-) -> Union[Function, LocalEntrypoint, MethodReference]:
-    """Using only an app - automatically infer a single "runnable" for a `modal run` invocation
+    Runnables includes all Functions, (class) Methods and Local Entrypoints, including web endpoints.
```

Comment:
Why are web endpoints considered "runnable"?

Reply:
`modal shell` can "run" them - happy to consider other names (but I'm only using this internally at the moment)

Reply:
Ah interesting, yeah I guess that makes sense.